### ISSUE TITLE
.pixi 4.0

### ISSUE THREAD

**Issue by CPKreuz** [PixiEditor/PixiParser#12]
The last few iterations of the .pixi file format have prioritized file size over everything else

This resulted in a few issues like:

- Not being able to tell if the file is an actual .pixi from the first few bytes as there is no magic number, resulting in a huge memory allocaction in some cases
- Hard to extend the format as the file version is in the middle of the file

# .pixi 4.0 Concept

The issues mentioned above will be fixed by .pixi 4.0

The foundation of this version should stay for the next versions to come

### Old "Header"

The pre 4.0 "header" consistet of a 4 byte long int32 representing the length of the document

A byte array of this size was then created, meaning that the first 4 bytes could result in a huge memory allocation mentioned before

### Concept Header

The header should cointaing the most important information for the parser so that it knows how to parse the file correctly

The order and length of the following properties should not change over any version

#### Magic Number

The new header should start with a magic number, something like

`20 50 49 58 49` representing `·PIXI` in ASCII

should be enough

If these are not the first bytes PixiParser will aready know this is (probably) not a .pixi file and can just throw a exception indicating that it's a corrupt file or another file format

#### File Version

After the magic number the version of the file should come

This could come in the form of a simple integer

#### Length

A int32 representing the total length of the document excluding the header

## Document Data

The document can be stored in a simple message pack again, starting with the size in the form of an int32 again

Basically just like the previous verision

The msg pack object could look like this

```
- Document
| - Width & Height
| - Swatches
----
| - Layer References
| | - Reference Type (e.g. Bitmap)
| | - Index
----
| - Bitmap Layers
| | Name
| | Width, Height & Offset
| | Opacity & Visibility
| | Layer Data offset
| | Layer Data length
----
| - Groups
| | Start & End Layer Reference Index
| | Name
| | Opacity & Visibility
----
| - Reference Layer
```

## Layer Data

The layer data is just a bunch of data reference from the document.

It can contain PNG's or other types of data

# Compression

Compression could be added to the Document and Layer data

**Comment by flabbet** [comment]
All good, although this need to include a precise byte decoding doc because otherwise no custom parser without a message pack could be written

**Comment by CPKreuz** [comment]
Implemented here
https://github.com/PixiEditor/PixiParser/pull/18
